tools/build/mingw_toolchain/file_collector.cpp

```diff
@@ -0,0 +1,218 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <direct.h>
+#include <iostream>
+#include <fstream>
+#include <shlwapi.h>
+#include <sstream>
+#include <string>
+#include <vector>
+#include <windows.h>
+
+#include "file_collector.h"
+
+namespace {
+
+std::string GetCWD() {
+  char buffer[1024];
+  _getcwd(buffer, sizeof(buffer));
+  return R"(\\?\)" + std::string(buffer) + '\\';
+}
+
+// Attemps to create a new folder in the temporary folder based off the PID.
+std::string GetTempDir(const char* baseName) {
+  TCHAR buffer[1024];
+  int len = GetTempPath(sizeof(buffer), buffer);
+
+  for (int i = 0, suffix = GetCurrentProcessId(); i < 10000; i++, suffix++) {
+    snprintf(buffer + len, sizeof(buffer) - len, "%s_%04x\\", baseName, suffix % 0x10000);
+    if (CreateDirectory(buffer, nullptr)) {
+      return std::string(buffer);
+    } else if (GetLastError() != ERROR_ALREADY_EXISTS) {
+      std::cerr << "Failed to create temporary directory " << buffer << std::endl;
+      abort();
+    }
+  }
+
+  std::cerr << "Couldn't find a unique temp dir" << std::endl;
+  abort();
+  return "";
+}
+
+// Copies source to target, optinally munging the target file name if
+// a file with the same name already exists at that location.
+std::string Copy(const std::string& source, const std::string& target) {
+  std::string result = target;
+  if (PathFileExists(target.c_str())) {
+    std::string::size_type p = target.rfind(".");
+    std::string base;
+    if (p == std::string::npos) {
+      base = target;
+    } else {
+      base = target.substr(0, p);
+    }
+
+    int s = 1;
+    while (true) {
+      std::stringstream ss;
+      ss << base << '_' << s++;
+      if (p != std::string::npos) {
+        ss << target.substr(p);
+      }
+      result = ss.str();
+      if (!PathFileExists(result.c_str())) {
+        break;
+      }
+    }
+  }
+
+  if (!CopyFile(source.c_str(), result.c_str(), true)) {
+    std::cerr << "Failed to copy file " << source << " to " << result << std::endl;
+    abort();
+  }
+
+  // Make the temporary files writtable, so they can be deleted later.
+  int attr = GetFileAttributes(result.c_str());
+  if (attr != INVALID_FILE_ATTRIBUTES && (attr & FILE_ATTRIBUTE_READONLY)) {
+    SetFileAttributes(result.c_str(), attr & ~FILE_ATTRIBUTE_READONLY);
+  }
+  return result;
+}
+
+} // namespace
+
+FileCollector::FileCollector(const std::string& baseName) :
+  baseName_(baseName),
+  cwd_(GetCWD()),
+  tmpDir_(""),
+  counter_(0) {
+}
+
+FileCollector::~FileCollector() {
+}
```

Comment:
Automatically clean up in destructor?

Reply:
I didn't do that, because I want it to leave the temp files around if there is an error, so it can be debugged.